core/network/fabric-interop-cc/contracts/interop/membership.go

```diff
@@ -58,16 +280,135 @@ func (s *SmartContract) CreateMembership(ctx contractapi.TransactionContextInter
+// TODO: Remove call to 'updateMembership' after creating Corda IIN Agents.
+func (s *SmartContract) UpdateMembership(ctx contractapi.TransactionContextInterface, counterAttestedMembershipSerialized string) error {
+	// Check if the caller has IIN agent privileges
+	if isIINAgent, err := wutils.IsClientIINAgent(ctx); err != nil {
+		return fmt.Errorf("IIN Agent client check error: %s", err)
+	} else if !isIINAgent {
+		// Check if the caller has network admin privileges
+		if isAdmin, err := wutils.IsClientNetworkAdmin(ctx); err != nil {
+			return fmt.Errorf("Admin client check error: %s", err)
+		} else if !isAdmin {
+			return fmt.Errorf("Caller neither a network admin nor an IIN Agent; access denied")
+		}
+		return updateMembership(s, ctx, counterAttestedMembershipSerialized)		// HACK to handle unattested memberships (for Corda) for backward compatibility
+	}
+
+	// Check if caller is one of the authorized IIN agents
+	callerId := ctx.GetClientIdentity()
+	callerCert, err := callerId.GetX509Certificate()
+	if err != nil {
+		return err
+	}
+	callerMemberUnit, err := callerId.GetMSPID()
+	if err != nil {
+		return err
+	}
+	err = verifyMemberInSecurityDomain1(s, ctx, "", callerCert, membershipLocalSecurityDomain, callerMemberUnit)
+	if err != nil {
+		return fmt.Errorf("Caller with identity %+v is not a designated IIN Agent: %+v", callerId, err)
+	}
+
+	counterAttestedMembership, err := decodeCounterAttestedMembership(counterAttestedMembershipSerialized)
+	if err != nil {
+		return fmt.Errorf("Counter Attested Membership Unmarshal error: %s", err)
+	}
+	decodedAttestedMembershipSet, err := base64.StdEncoding.DecodeString(counterAttestedMembership.AttestedMembershipSet)
+	if err != nil {
+		return fmt.Errorf("Attested membership set could not be decoded from base64: %s", err.Error())
+	}
+	var attestedMembershipSet identity.CounterAttestedMembership_AttestedMembershipSet
+	err = protoV2.Unmarshal(decodedAttestedMembershipSet, &attestedMembershipSet)
+	if err != nil {
+		return fmt.Errorf("Unable to unmarshal attested membership set: %s", err.Error())
+	}
+	decodedForeignMembership, err := base64.StdEncoding.DecodeString(attestedMembershipSet.Membership)
+	if err != nil {
+		return fmt.Errorf("Foreign membership could not be decoded from base64: %s", err.Error())
+	}
+	var foreignMembership common.Membership
+	err = protoV2.Unmarshal(decodedForeignMembership, &foreignMembership)
+	if err != nil {
+		return fmt.Errorf("Unable to unmarshal membership: %s", err.Error())
+	}
+
+	// Check presence of membership on ledger first
+	membershipKey, err := ctx.GetStub().CreateCompositeKey(membershipObjectType, []string{foreignMembership.SecurityDomain})
+	_, getErr := s.GetMembershipBySecurityDomain(ctx, foreignMembership.SecurityDomain)
+	if getErr != nil {
+		return getErr
+	}
+
+	// Match nonces across all attestations, local and foreign
```

Comment:
From this line to line 390, isn't it the same code as in `CreateMembership`, can they be moved to a separate helper function? say `validateCounterAttestedMembership`?